File: test_custom/cgi_bin/form_handler.py
#!/usr/bin/env python3
import os
import sys
import html
from urllib.parse import parse_qs

method = os.environ.get("REQUEST_METHOD", "GET")
content_type = os.environ.get("CONTENT_TYPE", "")

body = ""
if method == "POST":
    content_length = int(os.environ.get("CONTENT_LENGTH", 0))
    raw = sys.stdin.read(content_length)
    if "application/x-www-form-urlencoded" in content_type:
        parsed = parse_qs(raw)
        parts = []
        for k, v in parsed.items():
            parts.append(f"<tr><td>{html.escape(k)}</td><td>{html.escape(v[0])}</td></tr>")
        body = "".join(parts)
    else:
        body = f"<tr><td colspan='2'><pre>{html.escape(raw)}</pre></td></tr>"
else:
    qs = os.environ.get("QUERY_STRING", "")
    parsed = parse_qs(qs)
    parts = []
    for k, v in parsed.items():
        parts.append(f"<tr><td>{html.escape(k)}</td><td>{html.escape(v[0])}</td></tr>")
    body = "".join(parts) if parts else "<tr><td colspan='2'>No data</td></tr>"

sys.stdout.write("Content-Type: text/html\r\n\r\n")
sys.stdout.write("<html><head><title>Form Handler</title>\r\n")
sys.stdout.write("<style>body{font-family:monospace;margin:20px}table{border-collapse:collapse}th,td{border:1px solid #ddd;padding:8px}th{background:#2196F3;color:white}</style>\r\n")
sys.stdout.write("</head><body>\r\n")
sys.stdout.write(f"<h1>Form Data (Method: {html.escape(method)})</h1>\r\n")
sys.stdout.write("<table><tr><th>Field</th><th>Value</th></tr>\r\n")
sys.stdout.write(body + "\r\n")
sys.stdout.write("</table>\r\n")
sys.stdout.write("<p><a href='/cgi/form_handler.py'>Back</a></p>\r\n")
sys.stdout.write("</body></html>\r\n")

File: test_custom/cgi_bin/php/form_handler.php
<?php
$method = $_SERVER['REQUEST_METHOD'];
$data = [];

if ($method === 'POST') {
    $data = array_merge($_POST, $_GET);
} else {
    $data = $_GET;
}

echo "<html><head><title>PHP Form Handler</title>";
echo "<style>body{font-family:monospace;margin:20px}table{border-collapse:collapse}th,td{border:1px solid #ddd;padding:8px}th{background:#9C27B0;color:white}</style>";
echo "</head><body>";
echo "<h1>PHP Form Data (Method: " . htmlspecialchars($method) . ")</h1>";
echo "<table><tr><th>Field</th><th>Value</th></tr>";
if (!empty($data)) {
    foreach ($data as $key => $value) {
        echo "<tr><td>" . htmlspecialchars($key) . "</td><td>" . htmlspecialchars($value) . "</td></tr>";
    }
} else {
    echo "<tr><td colspan='2'>No data</td></tr>";
}
echo "</table>";
echo "<p><a href='/php/form_handler.php'>Back</a></p>";
echo "</body></html>";
?>
